### PR TITLE
MM-44842: Add restore user group command

### PR DESCRIPTION
**Pull request by plant99**
#### Summary

Add command to restore archived custom user group. 

Server-PR: https://github.com/mattermost/mattermost-server/pull/20369

**Note**: This PR also updates server version to a recent commit, thus the vendor changes and client refactorings.
#### Ticket Link

https://mattermost.atlassian.net/browse/MM-44842

### CODE REVIEW

**Comment by plant99** [comment]
@isacikgoz please review https://github.com/mattermost/mmctl/pull/614/commits/ddf019cbaf93320e5c26e5a1a014845a4fc16727 especially.

There were recent changes to dependencies in `master`, to resolve merge conflicts here's what I did

 - git pull origin master
 - fix conflicts in go.mod
 - go mod tidy
 - fix conflicts in `vendors/modules.txt`
 - go mod vendor

There was still an extra file in vendor, is there any next steps for me?

**Comment by BenCookie95** [comment]
@plant99 are you sure something didn't go wrong when you merged in master? 785 changes doesn't seem right

**Comment by plant99** [comment]
> @plant99 are you sure something didn't go wrong when you merged in master? 785 changes doesn't seem right

@BenCookie95 I think we're fine, since dependencies were updated, they were vendored in by Ibrahim in this commit: https://github.com/mattermost/mmctl/pull/614/commits/fd8e4da7e5f29031439e9c4a3fead6d832809f0a

File: commands/group.go [hunk 1]
```diff
@@ -308,3 +329,16 @@ func teamGroupListCmdF(c client.Client, cmd *cobra.Command, args []string) error
...
+	groupID := args[0]
+	_, resp, err := c.RestoreGroup(groupID, "")
+
+	if err != nil {
+		return err
+	}
+	if resp.StatusCode == http.StatusOK {
+		printer.Print("Group successfully restored with ID: " + groupID)
+	}
+	return nil
```
Comment (isacikgoz) [inline commands/group.go:343]: minor styling suggestion:
```suggestion
	if err != nil {
		return err
	}

	if resp.StatusCode == http.StatusOK {
		printer.Print("Group successfully restored with ID: " + groupID)
	}

	return nil
```

Comment (isacikgoz) [reply commands/group.go:343]: Additionally, is there any case that server returns `http.StatusOK` when there is no `error`?

File: commands/group_e2e_test.go [hunk 2]
```diff
@@ -507,3 +507,44 @@ func createTestGroupTeam(s *MmctlE2ETestSuite) (*model.Team, *model.Group, func(
...
+
+	defer func() {
+		_, err := s.th.App.DeleteGroup(group.Id)
+		s.Require().Nil(err)
+	}()
+
+	s.Run("Should allow group restore after deletion", func() {
+		_, appErr := s.th.App.DeleteGroup(group.Id)
+		s.Require().Nil(appErr)
+		printer.Clean()
```
Comment (isacikgoz) [inline commands/group_e2e_test.go:534]: nit:
```suggestion
		printer.Clean()

		_, appErr := s.th.App.DeleteGroup(group.Id)
		s.Require().Nil(appErr)
```

File: commands/group_e2e_test.go [hunk 3]
```diff
@@ -507,3 +507,44 @@ func createTestGroupTeam(s *MmctlE2ETestSuite) (*model.Team, *model.Group, func(
...
+	})
+	s.Require().Nil(appErr)
+	s.th.App.Srv().SetLicense(model.NewTestLicenseSKU(model.LicenseShortSkuProfessional, "ldap"))
+
+	defer func() {
+		_, err := s.th.App.DeleteGroup(group.Id)
+		s.Require().Nil(err)
+	}()
+
+	s.Run("Should allow group restore after deletion", func() {
```
Comment (isacikgoz) [inline commands/group_e2e_test.go:531]: Do we allow for all client types (eg. user, admin and local user) to restore groups? I couldn't see a test regarding to authorization.

File: commands/group_test.go [hunk 4]
```diff
@@ -1540,3 +1541,37 @@ func (s *MmctlUnitTestSuite) TestChannelGroupDisableCmdF() {
...
+
+		cmd := &cobra.Command{}
+		err := userGroupRestoreCmdF(s.client, cmd, []string{"groupId"})
+
+		s.Require().Nil(err)
+		s.Require().Len(printer.GetLines(), 1)
+	})
+
+	s.Run("User group restore command restores deleted user group", func() {
+		printer.Clean()
```
Comment (isacikgoz) [inline commands/group_test.go:1563]:
```suggestion
		printer.Clean()

```

**Review by isacikgoz — COMMENTED** [review]
LGTM overall. Have a couple of suggestions and a question.

File: commands/group.go (hunk 1 above)
Comment (plant99) [reply commands/group.go]: Thanks for pointing it out, https://github.com/mattermost/mattermost-server/pull/20369/files#diff-060eb667249b07159dcd89b520fa51777abf5bd355241d2e5c7e6bb58a93d367R1178 it only returns when RestoreGroup succeeds.

File: commands/group_e2e_test.go (hunk 3 above)
Comment (plant99) [reply commands/group_e2e_test.go:531]: https://github.com/mattermost/mmctl/pull/614/commits/5933073bb37fab78a838f709c4fd6226ec6633f7 @isacikgoz I've added a test, but it looks like in my naivety, I didn't create a role `PermissionRestoreCustomGroup` in https://github.com/mattermost/mattermost-server/pull/20369/files this PR.

I will create a PR to fix that, but do you think we should wait for that to merge this? Or can I just update this code when I create a new permission to restore groups.

**Review by isacikgoz — APPROVED** [review]
LGTM 👍

**Review by furqanmlk — APPROVED** [review]
Thanks @plant99 

Working as expected
1. Archive User Group

image

image


2. Restore using mmctl

image

3. User Group is restored 

image

**Pull request merged**